components/esp_modem/test/target_ota/components/manual_ota/manual_ota.cpp

```diff
@@ -0,0 +1,259 @@
+                if (err != ESP_OK) {
+                    ESP_LOGE(TAG, "esp_ota_begin failed (%s)", esp_err_to_name(err));
+                    esp_ota_abort(update_handle_);
+                    return fail_cleanup();
+                }
+                ESP_LOGI(TAG, "esp_ota_begin succeeded");
+            } else {
+                ESP_LOGE(TAG, "Received chunk doesn't contain app descriptor");
+                esp_ota_abort(update_handle_);
+                return fail_cleanup();
+            }
+        }
+        err = esp_ota_write(update_handle_, (const void *)buffer_.data(), data_read);
+        if (err != ESP_OK) {
+            esp_ota_abort(update_handle_);
+            return fail_cleanup();
+        }
+        file_length_ += data_read;
+        ESP_LOGI(TAG, "Written image length %d", file_length_);
+
+        if (image_length_ == file_length_) {
+            status = state::END;
+            return false;
+        }
+
+        if (!prepare_reconnect()) {
+            esp_ota_abort(update_handle_);
+            return fail_cleanup();
+        }
+
+    } else if (data_read == 0) {
+        if (file_length_ == 0) {
+            // Try to handle possible HTTP redirections
+            int status_code = esp_http_client_get_status_code(http_);
+            ESP_LOGW(TAG, "Status code: %d", status_code);
+            err = esp_http_client_set_redirection(http_);
+            if (err != ESP_OK) {
+                ESP_LOGE(TAG, "URL redirection Failed");
+                esp_ota_abort(update_handle_);
+                return fail_cleanup();
+            }
+
+            err = esp_http_client_open(http_, 0);
+            if (err != ESP_OK) {
+                ESP_LOGE(TAG, "Failed to open HTTP connection: %s", esp_err_to_name(err));
+                return fail_cleanup();
+            }
+            esp_http_client_fetch_headers(http_);
+        }
+    }
+
+    return true;
+}
+
+bool manual_ota::prepare_reconnect()
+{
+    esp_http_client_set_method(http_, HTTP_METHOD_GET);
+    char *header_val = nullptr;
+    if ((image_length_ - file_length_) > max_buffer_size_) {
+        asprintf(&header_val, "bytes=%d-%d", file_length_, (file_length_ + max_buffer_size_ - 1));
+    } else {
+        asprintf(&header_val, "bytes=%d-", file_length_);
+    }
+    if (header_val == nullptr) {
+        ESP_LOGE(TAG, "Failed to allocate memory for HTTP header");
+        return false;
+    }
+    esp_http_client_set_header(http_, "Range", header_val);
+    free(header_val);
+    return true;
+}
+
+bool manual_ota::fail_cleanup()
+{
+    esp_http_client_close(http_);
+    esp_http_client_cleanup(http_);
+    status = state::FAIL;
+    return false;
+}
+
+bool manual_ota::end()
+{
+    if (status == state::END) {
+        if (!esp_http_client_is_complete_data_received(http_)) {
+            ESP_LOGE(TAG, "Error in receiving complete file");
+            return fail_cleanup();
```

Comment:
Would it make sense to add an `err` label in this function and use `goto` on failure?

Reply:
consider wrapping `http_client` into a class and perform the cleanup on destruction.

Reply:
wrapped to an internal class and reworked to cleanup in destructors